### ISSUE TITLE
Docker 1.12 swarm mode multi-manager setup does not tolerate node failures

### ISSUE THREAD

**Issue by ginkel** [docker/docker#24329]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:        Sat Jul  2 00:15:19 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:        Sat Jul  2 00:15:19 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 10
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: active
 NodeID: bqb7y2x9yixyfvlm1cba1ackz
 IsManager: Yes
 Managers: 3
 Nodes: 3
 CACertHash: sha256:1f0ee0ef3bda17565d3a6cc9690ea3ae059a3819945dd864656b277c9068c77a
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.469 GiB
Name: buddy-docker-02
ID: NGBD:3EYW:26LC:RIL7:JNF3:Z2E6:ZS7H:A56E:XDAH:PFHM:JF4V:L3VV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Virtualized sandbox environment based on Vagrant / debian/jessie64.

**Steps to reproduce the issue:**
1. Install Docker 1.12 on three nodes (n01, n02, n03).
2. Initialize the swarm on n01: `docker swarm init --listen-addr <n01_ip>:2377 --auto-accept worker --auto-accept manager`
3. On n2, join the cluster: `docker swarm join --manager --listen-addr <n02_ip>:2377 <n01_ip>:2377`
4. On n3, join the cluster: `docker swarm join --manager --listen-addr <n03_ip>:2377 <n02_ip>:2377`
5. Create a service: `docker service create --constraint ‘affinity:container!=nginx*’ --replicas 3 --name nginx -p 80:80 nginx`
6. 
```
$ docker service ls 
ID            NAME   REPLICAS  IMAGE  COMMAND
bu0mnmlqifiv  nginx  3/3       nginx

$ docker node ls
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
2ga9bs069m8h7808iizq3fym2 *  n01       Accepted    Ready   Active        Reachable
8l9b2q8l6xtwuqjcfcb43fg2p    n03       Accepted    Ready   Active        Leader
bqb7y2x9yixyfvlm1cba1ackz    n02       Accepted    Ready   Active        Reachable
```
6. Kill n01

7. On n02 and n03:

```
docker node ls
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
2ga9bs069m8h7808iizq3fym2    n01       Accepted    Down    Active        Reachable
8l9b2q8l6xtwuqjcfcb43fg2p *  n03       Accepted    Ready   Active        Reachable
bqb7y2x9yixyfvlm1cba1ackz    n02       Accepted    Down    Active        Leader
```

I.e., the leader is marked as down (although it isn't), but the remaining manager node marked as active also does not get promoted leader.

Also, all of a sudden, the cluster has become inconsistent and four replicas (instead of three) are running  in the cluster:

On n02:

```
$ docker service ls 
ID            NAME   REPLICAS  IMAGE  COMMAND
bu0mnmlqifiv  nginx  3/3       nginx 

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
a346240c79a6        nginx:latest        "nginx -g 'daemon off"   6 minutes ago       Up 6 minutes        80/tcp, 443/tcp     nginx.3.2bpe57wto47t4swnjx4qkqmsb
```

On n03:

```
$ docker service ls 
ID            NAME   REPLICAS  IMAGE  COMMAND
bu0mnmlqifiv  nginx  3/3       nginx  

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
42c8f53a48c7        nginx:latest        "nginx -g 'daemon off"   4 minutes ago       Up 4 minutes        80/tcp, 443/tcp     nginx.1.f2mbb04jmwhkh2vd6nagtjnjp
6358044e79d5        nginx:latest        "nginx -g 'daemon off"   4 minutes ago       Up 4 minutes        80/tcp, 443/tcp     nginx.3.4v5f1z4bup06bxp8e9yj41mqc
7e1e7e955bb7        nginx:latest        "nginx -g 'daemon off"   8 minutes ago       Up 8 minutes        80/tcp, 443/tcp     nginx.2.b483sq8h2whum23se5gx3tv8y
```

**Describe the results you received:**

See above.

**Describe the results you expected:**

I'd expect the cluster to remain consistent, i.e., the two remaining manager nodes should reach a consensus and re-schedule the containers. Both n02 and n03 should show as up as active, one of them should have been elected as leader.

**Additional information you deem important (e.g. issue happens only occasionally):**

The above setup has been created using a kind of manager daisy chain, i.e., n02 joined n01, n03 joined n02. The problem also occurs if all followers are initially joining n01.

**Comment by aaronlehmann** [comment]
ping @abronan @LK4D4

**Comment by cohenaj194** [comment]
I have another issue that was created after using the same exact steps @ginkel outlined in **Steps to reproduce the issue:**

After failing my `n01` node with `sudo service docker stop` the node `n01` was then removed from the output of `docker node ls` however it continues to function properly as a swarm manager.  Also note that my configuration is slightly different as I am using 3 managers (whose availability is drained) and 3 workers.

**Comment by abronan** [comment]
@ginkel I can't seem to reproduce using `master` or `1.12.0-rc3`. I'm following the same scenario described above and I get (on `n2`, `n3`):

```
~# docker node ls
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
0ff2237si59mivnrbmgknzn7h    node02    Accepted    Ready   Active        Reachable
3dlwrztkpolbyzq4473jlfnmt    node01    Accepted    Down    Active        Unreachable
d68wciofy5rgarwb695l2pnvl *  node03    Accepted    Ready   Active        Leader
```

Also, `docker service ls` as well as `docker ps` correctly reports 3 containers and not 4. Noting that `service ls` reports `2/3` for a few seconds and then reconcile to get `3/3`.

Can you systematically reproduce that scenario (especially the Leader being marked as down)? Would you also mind retrying on `master` to see if that solves the issue? Thanks.

A bit more details: Losing the quorum (2 nodes reported as down on a 3 nodes Manager cluster), will bring the cluster to an inconsistent state until a majority of nodes (in your case **2**) recovers and the reconciliation loop figures out what to do (to rebalance and schedule the right amount of tasks). That is why it is recommended to use at least *5 managers* to avoid falling into this case.

**Comment by ginkel** [comment]
@abronan Yes, I can easily reproduce the problem (on a system set up from scratch using Vagrant). I'm not sure whether there's anything weird with my Vagrant/VirtualBox setup that serves as a testbed. There might be one special thing: The VMs have two ethernet interfaces: eth0 (NAT) and eth1 (host-only, for node-to-node communication). I'm binding the swarm to eth1 of each node, respectively, so I wouldn't expect any trouble, though...

When n01 goes down, this is all I see in the n02 syslog:

```
Jul  6 04:50:13 jessie dockerd[18040]: time="2016-07-06T04:50:13.833901768Z" level=info msg="2016/07/06 04:50:13 [INFO] memberlist: Suspect n01 has failed, no acks received\n"
Jul  6 04:50:14 jessie dockerd[18040]: time="2016-07-06T04:50:14.834208676Z" level=info msg="2016/07/06 04:50:14 [INFO] memberlist: Suspect n01 has failed, no acks received\n"
Jul  6 04:50:16 jessie dockerd[18040]: time="2016-07-06T04:50:16.832184549Z" level=info msg="2016/07/06 04:50:16 [INFO] memberlist: Suspect n01 has failed, no acks received\n"
Jul  6 04:50:17 jessie dockerd[18040]: time="2016-07-06T04:50:17.315493373Z" level=info msg="2016/07/06 04:50:17 [INFO] memberlist: Marking n01 as failed, suspect timeout reached\n"
```

Would it help if I could share my Vagrantfile (+ Ansible playbook)?

Is there a Debian repo with packages built from master (or installation instructions how to install master from scratch)? That would really help getting the scenario reproduced on master.

Thanks for your help!

**Comment by ginkel** [comment]
I did a further test with five nodes: For the sake of simplicity, `n02` joined through `n01` and `n03`..`n05` joined through `n02`.

This is the cluster state after killing `n01` (observed on `n02`):

```
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
29ol9ym69g6fxouy4h6uu941h    n04       Accepted    Ready   Active        Reachable
5s33rmkcv6x8hd8v7n7lwelwj    n03       Accepted    Ready   Active        Reachable
bedho8pmgc8ff1z2fh5pkj3ta    n05       Accepted    Ready   Active        Leader
erwv19q5dhv3tw06u2eqbnt49 *  n02       Accepted    Down    Active        Reachable
esn2ngrge9xpa843qev8t1eyp    n01       Accepted    Down    Active        Reachable
```

To me this looks as if `n02` stays bound to `n01` as its upstream and does not attempt to locate a new manager as soon as `n01` is gone.

**Comment by thaJeztah** [comment]
Marking this a P1 until it's clear what is causing this

**Comment by ginkel** [comment]
I put my test setup into a Git repo: https://github.com/tgbyte/docker-issue-24329

Requires Ansible on the host (I used v2.0.0.2).

This is currently untested as I had to strip down my original version and am currently in a train with pretty bad WiFi, which makes running Vagrant somewhat tricky. ;-) I'll be back later today and will add another comment once I have verified the setup works.

**Comment by aaronlehmann** [comment]
I think the issue here is that when you take down `n01`, it's not closing the TCP connections between `n01` and the other nodes. We use the status of this connection to judge manager reachability and worker status, so it would explain the strange results in `node ls` like `n01` still being reachable.

The worker does not appear to put a timeout on its RPC call that sends heartbeats to the manager. So I believe this call is hanging because the manager is no longer responsive, but the TCP connection has not been torn down.

I will submit PRs to the swarmkit repository to:

1) Add a timeout on the worker-to-manager heartbeat, so reaching this timeout causes the worker to fail over to another manager.
2) Make raft clustering (used by managers) more intelligent about determining reachability.

Thanks for reporting this.

**Comment by thaJeztah** [comment]
https://github.com/docker/swarmkit/pull/1136 was merged, so I think this should be fixed upon the next SwarmKit vendoring update

**Comment by ginkel** [comment]
@aaronlehmann That sounds plausible as I killed `n01` by just turning off the VM the hard way, so TCP/IP connections certainly did not have a chance to be shut down in a controlled fashion.

Thanks for the speedy resolution! :-)

BTW, if there is a feasible way for me to re-test the fix before 1.12 is released, please let me know.

**Comment by tiborvass** [comment]
Fixed by #23312

**Comment by aaronlehmann** [comment]
Fixed by #24459.

**Comment by ginkel** [comment]
I can confirm that this is fixed as of rc4. Thanks for the speedy bugfix! :-)
